File: Core/Architect.API.Core.DataAccess/General/InternalTemplate.vb
Imports System.Configuration
Imports Architect.Common.Proxy

Namespace Architect.API.Core.DataAccess.General
    Public NotInheritable Class InternalTemplate

        Public Shared Function Retrieve(companyId As Integer, key As String) As Architect.API.Core.Contracts.General.InternalTemplate
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.General.InternalTemplate = New Contracts.General.InternalTemplate With {.Subject = "Plantilla no encontrada",
                                                                                                               .Body = "Plantilla no encontrada"}
            With New DataManagerFactory("SELECT Subject, Body, EmailTo, MasterTemplateId " &
                                          "FROM Template " &
                                         "WHERE Key=@:Key AND (CompanyId=@:CompanyId OR CompanyId=0) Order by CompanyId DESC", "Template", "Research")
                .AddParameter("Key", DbType.AnsiString, 40, False, key)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                If ConfigurationManager.AppSettings("Working.Mode") <> "Development" Then
                    .Cache = Common.Enumerations.EnumCache.CacheWithFullParameters
                End If
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = New Architect.API.Core.Contracts.General.InternalTemplate With {.Subject = rows.FirstRow.StringValue("Subject"),
                                                                                     .Body = rows.FirstRow.StringValue("Body"),
                                                                                     .EmailTo = rows.FirstRow.StringValue("EmailTo"),
                                                                                     .MasterTemplateId = rows.FirstRow.IntegerValue("MasterTemplateId")}
                End If
            End With
            Return result
        End Function

        Public Shared Function RetrieveById(companyId As Integer, id As Integer) As Architect.API.Core.Contracts.General.InternalTemplate
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.General.InternalTemplate = New Contracts.General.InternalTemplate With {.Subject = "Plantilla no encontrada",
                                                                                                               .Body = "Plantilla no encontrada"}
            With New DataManagerFactory("SELECT Subject, Body, EmailTo, MasterTemplateId " &
                                          "FROM Template " &
                                         "WHERE Id=@:Id AND (CompanyId=@:CompanyId OR CompanyId=0) Order by CompanyId DESC", "Template", "Research")
                .AddParameter("Id", DbType.Decimal, 9, False, id)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                If ConfigurationManager.AppSettings("Working.Mode") <> "Development" Then
                    .Cache = Common.Enumerations.EnumCache.CacheWithFullParameters
                End If
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = New Architect.API.Core.Contracts.General.InternalTemplate With {.Subject = rows.FirstRow.StringValue("Subject"),
                                                                                     .Body = rows.FirstRow.StringValue("Body"),
                                                                                     .EmailTo = rows.FirstRow.StringValue("EmailTo"),
                                                                                     .MasterTemplateId = rows.FirstRow.IntegerValue("MasterTemplateId")}
                End If
            End With
            Return result
        End Function

    End Class

End Namespace

File: Core/Architect.API.Core.DataAccess/General/InternalTemplate.cs
using Architect.DataFactory;
using Architect.Utilities.Extensions;
using System;
using System.Configuration;
using System.Data;
using DbType = Architect.DataFactory.Enumerations.DbType;

namespace Architect.API.Core.DataAccess.General
{
    public sealed class InternalTemplate
    {
        /// <summary>
        /// Recupera una plantilla por medio de la clave de la misma para la compañía en tratamiento o para la compañía 0.
        /// </summary>
        /// <param name="companyId"></param>
        /// <param name="key"></param>
        /// <returns></returns>
        public static Contracts.General.InternalTemplate Retrieve(int companyId, string key)
        {
            Contracts.General.InternalTemplate result = new Contracts.General.InternalTemplate()
            {
                Subject = "Plantilla no encontrada",
                Body = "Plantilla no encontrada"
            };

            Database.Select(@"SELECT Subject, Body, EmailTo, MasterTemplateId
                               FROM Template
                              WHERE Key=:Key AND (CompanyId=:CompanyId OR CompanyId=0) ORDER BY CompanyId DESC")
                        .AddParameter("Key", DbType.AnsiString, 40, key)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Cache(ConfigurationManager.AppSettings["Working.Mode"] != "Development")
                        .Query(null, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {

                            result = new Contracts.General.InternalTemplate()
                            {
                                Subject = reader.StringValue("Subject"),
                                Body = reader.StringValue("Body"),
                                EmailTo = reader.StringValue("EmailTo"),
                                MasterTemplateId = reader.IntegerValue("MasterTemplateId")
                            };

                        }), false);

            return result;
        }

        public static Contracts.General.InternalTemplate RetrieveById(int companyId, int id)
        {
            Contracts.General.InternalTemplate result = new Contracts.General.InternalTemplate()
            {
                Subject = "Plantilla no encontrada",
                Body = "Plantilla no encontrada"
            };

            Database.Select(@"SELECT Subject, Body, EmailTo, MasterTemplateId 
                               FROM Template 
                              WHERE Id=:Id AND (CompanyId=:CompanyId OR CompanyId=0) Order by CompanyId DESC")
                        .AddParameter("Id", DbType.Decimal, 9, id)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Cache(ConfigurationManager.AppSettings["Working.Mode"] != "Development")
                        .Query(null, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result = new Contracts.General.InternalTemplate()
                            {
                                Subject = reader.StringValue("Subject"),
                                Body = reader.StringValue("Body"),
                                EmailTo = reader.StringValue("EmailTo"),
                                MasterTemplateId = reader.IntegerValue("MasterTemplateId")
                            };

                        }), false);

            return result;
        }
    }
}